app/models/remote_execution_provider.rb

```diff
@@ -9,7 +9,7 @@ def provider_for(type)
 
     def registered_name
       klass = self
-      providers.key(klass)
+      ::RemoteExecutionProvider.providers.key(klass)
```

Comment:
Seems logical to me if we only allow registering all the providers in the scope of `RemoteExecutionProvider` only. I mean, any subclass of the `RemoteExecutionProvider` can have its own registry and to make `registered_name` work properly for the subclass, it needs to redefine `registered_name` to use `SubclassProvider` scope for the registry.

If it's by design that registration of the providers must be done via `RemoteExecutionProvider.register` only then ACK.

Reply:
Could this https://github.com/theforeman/foreman_remote_execution/pull/757/files#diff-5c51055144db11e106aa62a1537c9ca8e4cafce1c5503000eb0d8b7ce232b232R31 also fail then? Or any method that uses `providers`?

I mean, I'm not so familiar with the usage of subclasses. If they are more or less `constants only` then it's okay, but if all those methods can be called on a subclass, wouldn't it behave inconsistently?

Reply:
In theory it could, but the only place where we call `.provider_for` that I'm aware of is [here](https://github.com/theforeman/foreman_remote_execution/blob/master/app/models/job_template.rb#L132) and then the rest of the methods called on the returned provider *should* be safe, with the exception of `#proxy_features`.